oeps/oep-0033-bp-code-autoformatting.rst

```diff
@@ -0,0 +1,127 @@
+| `Review Period` | 2019-03-04 - 2019-04-01                                |
++-----------------+--------------------------------------------------------+
+
+Context
+-------
+
+In a large software project or ecosystem like the Open edX codebase, consistency
+of code formatting makes it easier to read the code. In an ecosystem with many
+contributors, maintaining that consistency requires automated tools (such as
+linters), consistent code review, or both. In general, minimizing the number
+of review comments that a reviewer has to make on an incoming change makes
+review easier, and lessens the number of back-and-forths required to merge
+the code. As such, linters that can be run locally should be preferred.
+Similarly, ambiguity in style guidelines should be minimized, so that there
+are fewer places for commentary around what amounts to personal preference
+(which might then instigate unnecessary debate).
+
+In short, more consistent, less ambiguous standards for code formatting
+make for more productive code reviews and waste less time on inconsequential
+details.
+
+Decision
+--------
+
+We will select automatic code formatters for all of our major languages. These
+formatters will have both a lint-mode (that will be run in continuous
+integration to ensure consistent formatting) and an autoformat mode (which
+can be integrated with editors to free developers from having to think about
+code formatting). These formatting standards will be applied across all
+Open edX-authored repositories.
+
+Selected tools should, ideally, require minimal configuration. Any required
+configuration should be added to `edx/edx-lint`_ in order to allow it to
+be centralized and standardized across repositories in the Open edX ecosystem.
+
+The current recommended formatters are:
+
+    - Python: `black`_
+    - Javascript: `prettier`_
+    - CSS/SASS: `prettier`_
+    - HTML: `prettier`_
+    - JSON: `prettier`_
+    - PHP: `prettier-php`_
+
+.. _black: https://github.com/ambv/black
+.. _prettier: https://prettier.io/
+.. _prettier-php: https://github.com/prettier/plugin-php
+.. _`edx/edx-lint`: https://github.com/edx/edx-lint
+
+Formatting should exposed via the standard ``Makefile`` commands
+``format`` (which should autoformat all code in the repository)
+and ``check-format`` (which should validate that all code in the
+repository is formatted). ``check-format`` should also be called
+as part of ``make quality``.
+
+When bulk-formatting the entire repo for the first time, all changes should
+be done in as few commits as possible (ideally, only one). Then, the
+commit hashes should be added to a ``.git-blame-ignore-revs`` file.
+This will allow the `git hyper-blame`_ tool to ignore them by default,
+to preserve the easy ability to see who changed particular file lines.
+
+.. _git hyper-blame: https://commondatastorage.googleapis.com/chrome-infra-docs/flat/depot_tools/docs/html/git-hyper-blame.html
```

Comment:
While this seems interesting, I don't think it helps for many use cases.
Does Github suport it?
I know that my text editor integration (`vim-fugitive`) explicitly doesn't support it now [1].
I'm worried the impact here would be small.

- [1] https://github.com/tpope/vim-fugitive/issues/1058

Reply:
Yeah, Github doesn't support it. I'm willing to include it if people want it (or even if it would be marginally useful), because it isn't a big lift to add (but I agree, it's also doesn't solve a lot of use-cases).

Reply:
Yeah, I guess I'm not saying _don't_ mention it, as much as I wanted to be clear that it shouldn't be interpreted as end-all solution.